analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/DeconstructionAssignment.cs

```diff
@@ -0,0 +1,89 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn.OperationProcessors;
+
+internal sealed class DeconstructionAssignment : SimpleProcessor<IDeconstructionAssignmentOperationWrapper>
+{
+    protected override IDeconstructionAssignmentOperationWrapper Convert(IOperation operation) =>
+        IDeconstructionAssignmentOperationWrapper.FromOperation(operation);
+
+    protected override ProgramState Process(SymbolicContext context, IDeconstructionAssignmentOperationWrapper assignment)
+    {
+        var operationValues = CollectOperationValues(context.State, assignment.Target, assignment.Value);
+        var newState = context.State;
+        foreach (var (tupleMember, value) in operationValues)
+        {
+            newState = newState.SetOperationAndSymbolValue(tupleMember, value);
+        }
+        return newState;
+    }
+
+    private static IEnumerable<OperationValue> CollectOperationValues(ProgramState state, IOperation target, IOperation value)
+    {
+        var operationValues = new List<OperationValue>();
+        var leftTupleElements = TupleElements(target, state);
+        // If the right side is a tuple, then every symbol/constraint is copied to the left side.
+        if (Unwrap(value, state).AsTuple() is { } rightSideTuple)
+        {
+            var rightTupleElements = TupleElements(rightSideTuple.WrappedOperation, state);
+            for (var i = 0; i < leftTupleElements.Length; i++)
+            {
+                var leftTupleMember = leftTupleElements[i];
+                var rightTupleMember = rightTupleElements[i];
+                if (leftTupleMember.Kind == OperationKindEx.Discard)
+                {
+                    continue;
+                }
+                if (leftTupleMember.AsTuple() is { } nestedTuple)
+                {
+                    operationValues.AddRange(CollectOperationValues(state, nestedTuple.WrappedOperation, rightTupleMember));
+                }
+                else
+                {
+                    operationValues.Add(new(leftTupleMember, state[rightTupleMember]));
+                }
```

Comment:
This could be rewritten using something like this (with better formatting):
```
leftTupleMembers.Where(x => x.Kind != OperationKindEx.Discard)
    .SelectMany((x, i) => leftTuplemember.AsTuple() is { } nestedTuple
        ? CollectOperationValues(state, nestedTuple.WrappedOperation, rightTupleElements[i])
        : new[] { new(leftTupleMember, state[rightTupleMember]) }
```
This is optional, though. Going imperatively here is fine.

Reply:
I find the imperative version more readable. `Where(x => x.Kind != OperationKindEx.Discard)` could only be skipped at the end (because of the indexing).